### ISSUE TITLE
OM autonomously creates integers from reals

### ISSUE THREAD

**Issue by max-privato** [OpenModelica/OpenModelica#8062]
### Description
When using PowerGrids models, which allow displaying powers in MVA, one can enter them in MVA, using real numbers (having the dot). OM inserts them in the code as integers, then complains that integer numbers are used instead of real's!

### Steps to Reproduce
- Create a new model
- drag on it a component  Electrical.Buses.EquivalentGrid from PowerGrids library
- change SNom into 2500.0 MVA (using the MVA unit from the drop-down menu)
-=> OM puts into code 2500000000 and immediately complains that it is an unacceptable integer and that it will be converted into a real. 
Here I see two issues:
- the real number is unduly converted into an integer
- the writing is unacceptably long. I mean writing 2500000000. or 2500000000.0 would be unacceptable as well. Much better would be 2500.0e6 or 2.5e9

### Expected Behavior
1) the real number is kept as a real
2) the writing is reasonable readable, i.e.e has not so many, many unuseful digits displayed

### Version and OS

OM 1.18.0 WIn 10 64 bit
 - OpenModelica Version: [e.g. `omc --version` or `Help->About OMEdit` from OMEdit]
 - OS: [e.g. Windows 10, 64 bit]
 - Versions of used Modelica libraries if applicable

**Comment by max-privato** [comment]
If to reduce the complexity, we should relax some requirement, this could be, for me, getting back exactly our text. 
I mean, if I enter a power of 2500 MVA, I would not be too disappointed to get back 2.5e9 VA. or 2500.0 MVA.
Much worse are the two issues I listed in the ticket's description. 

This for ordinary quantities. For degrees, the situation more involute.

**Comment by sjoelund** [comment]
The problem with putting 2500 MVA into the code is that MSL is SI-units only, and changing units in the code doesn't work.
Regardless, only reals have a unit so it shouldn't be hard to force printing as a real in some way if there is a conversation factor.

**Comment by max-privato** [comment]
> The problem with putting 2500 MVA into the code is that MSL is SI-units only, 

Yes, sure. I intended that OMEdit, when the number 2.5e9 read from the mo, can show it in Electrical.Buses.EquivalentGrid as 2500 MVA, since its dialog box already has MVA as units of measure (SSC has Types,ApparentPower as type)

**Comment by sjoelund** [comment]
> > The problem with putting 2500 MVA into the code is that MSL is SI-units only,
> 
> Yes, sure. I intended that OMEdit, when the number 2.5e9 read from the mo, can show it in Electrical.Buses.EquivalentGrid as 2500 MVA, since its dialog box already has MVA as units of measure (SSC has Types,ApparentPower as type)

It would only know the unit if you put the displayUnit as MVA and it wouldn't know you input it as an integer and might display it as 2.5e3 MVA

**Comment by max-privato** [comment]
I agree: the simplest way for OMEdit is show 2.5e3 MVA. 
It would not be impossible to show 2500 MVA, but this requires some analysis on the number, similar to what is done by good plotting programs for axes: the best of them avoid showing e3 (or whatever exponent is needed))  on every number on tic marks.  Recently also OMEdit has reached this level, and therefore already knows how to do these things

**Comment by adeas31** [comment]
The actual value is always stored in unit regardless of whatever display unit it has. In this case the value is always stored in `V.A`. When you enter a value of `2500.0` in `MVA` then OMEdit call,

```
convertUnits("MVA","V.A")
(true,1e-06,0)
```

First it will check if the units are compatible and what offset and scale factor has to be used to convert the value.

> @adeas31, could you summarize the current logic used by OMEdit to handle these unit conversions?

The unit conversion is done as,

`(value - offset) / scaleFactor`

which in this case is,

`(2500.0 - 0) / 1e-06`

That gives the value of `2500000000`. Now if I use standard string conversion of this number using `QString::number(double value, char f =g, int prec = 6)` this gives `2.5e+09` but Martin wrote a custom conversion method that uses a precision of 16 which will of course give us back `2500000000`. So if OMC fails to parse `2500000000` then we should not have precision of 16 in OMEdit.

**Comment by max-privato** [comment]
> That gives the value of `2500000000`. Now if I use standard string conversion of this number using `QString::number(double value, char f =g, int prec = 6)` this gives `2.5e+09` but Martin wrote a custom conversion method that uses a precision of 16 which will of course give us back `2500000000`. So if OMC fails to parse `2500000000` then we should not have precision of 16 in OMEdit.

I don't want to enter into details of your number-converting code, but I think that being able to write on the code 2.5e09 instead of 2500000000.0 is a sign of good quality software. After all, the exponential form has been invented exactly to ease reading numbers; it is especially useful in cases like this.

**Comment by max-privato** [comment]
Update.
With 1.20.0 the situation has dramatically worsened.
Use the following procedure
- Create a new model
- drag on it a component Electrical.Buses.EquivalentGrid from PowerGrids library
- change SNom into 2500.0 MVA (using the MVA unit from the drop-down menu)
-=> OM puts into code -1794967296, which is plain wrong

**Comment by adeas31** [comment]
@perost does it have anything to do with the changes you did for `setElementModifierValue`?

```
loadString("model M
  PowerGrids.Electrical.Buses.EquivalentGrid equivalentGrid;
equation

annotation(
    uses(PowerGrids(version = \"1.0.2\")));
end M;");
getErrorString();
setElementModifierValue(M, equivalentGrid.SNom, $Code(=2500000000));
getErrorString();
list(M);
getErrorString();
```

produces,

```
true
""
Ok
""
"model M
  PowerGrids.Electrical.Buses.EquivalentGrid equivalentGrid(SNom = -1794967296);
equation

  annotation(
    uses(PowerGrids(version = \"1.0.2\")));
end M;"
""

```

`setElementModifierValue` sets the value to `2500000000` but the Modelica code shows `-1794967296`.

**Comment by perost** [comment]
> @perost does it have anything to do with the changes you did for `setElementModifierValue`?

No, the only thing I changed was to add a case for handling redeclares. However, your example works correctly for me, so I guess it might be a Windows issue.

The issue is obviously that the value is converted to a 32-bit signed integer at some point, which causes it to wrap around and become negative. But I'm not sure where that could happen.

**Comment by perost** [comment]
@adeas31: Does this work correctly for you?
```
loadString("
  model M
    Integer n = 2500000000;
  end M;
");
list(M);
```

**Comment by adeas31** [comment]
No,

```
true
"model M
  Integer n = -1794967296;
end M;"
```

**Comment by sjoelund** [comment]
Could also be toString/etc that do not use long long and associated function?

**Comment by perost** [comment]
> Could also be toString/etc that do not use long long and associated function?

Yes, that's perhaps a more likely issue, the parser looks like it should be fine.

**Comment by adeas31** [comment]
I ran the IntegerTest in #8691 and that works.

**Comment by perost** [comment]
The issue is with the builtin MetaModelica functions `intString/stringInt` which uses `long` instead of the appropriate 64-bit type. I've made a fix for it, but Github seems to be having some issues at the moment so I'll push it in later.

**Comment by perost** [comment]
The issue with the incorrect integer value should hopefully be fixed in #10453. I haven't actually tested it on Windows though, so let me know if it works or if I've broken everything.

**Comment by casella** [comment]
@adeas31 I hope @perost's fix with 64-bit ints works. However, that shouldn't have been necessary. I mean, `(2500.0 - 0) / 1e-06` is a Real expression, it should have returned `2500000000.0` (a Real), not `2500000000` (an Integer). That would have prevented the issue a priori. Why is an Integer value returned?

**Comment by casella** [comment]
@ceraolo, please check with the next nightly build.

**Comment by sjoelund** [comment]
> Why is an Integer value returned?

I guess because that's how Qt prints floating point numbers that are integers. So 2500.0 / 1e-6 it puts 2500000000 in the Qt widget and that text is then sent back to OMC, where it is parsed as an integer instead.

**Comment by casella** [comment]
@ceraolo I get the right value now, so @perost's fix worked. I guess we should fix this properly together with #8865, later this year.

**Comment by ceraolo** [comment]
Now the issue in https://github.com/OpenModelica/OpenModelica/issues/8062#issuecomment-1484031903 is solved (I checked under ubuntu).

The issue of this ticket's description is however still there: OM converts reals into integers, I don't kow why. I still hate seeing 2.5e9 being written 25000000000 (too many zeros and absence of dot).

**Comment by casella** [comment]
@ceraolo, this issue should be easily fixed once PR 11168 is merged in. At that point, we'll use a function to represent Real expressions that will only use decimal representation if there are at most 5 digits before the decimal point and at most 3 zeros before the decimal point, otherwise it will automatically switch to exponential form. For a discussion of the rationale, please check [this comment](https://github.com/OpenModelica/OpenModelica/issues/7465#issuecomment-1722285368).

**Comment by adeas31** [comment]
Fixed with #11346
